### PR TITLE
RISC-V: implement far branches

### PR DESCRIPTION
**Pull request by janvrany**

### CODE REVIEW

**Comment by janvrany** [comment]
@aviansie-ben Thank you very much for your help! My code is shameless copy of yours. I'd be nice to somehow share at least parts, but not sure how.

**Comment by janvrany** [comment]
Found a bug in `doBinaryEncoding()`, fixed in (force-pushed) [ca6296](https://github.com/eclipse/omr/commit/ca629699f603c6d6210bf221b1fbbe61f5e8653e)

File: compiler/riscv/codegen/RVInstruction.cpp [hunk 1]
```diff
@@ -411,6 +412,52 @@ uint8_t *TR::BtypeInstruction::generateBinaryEncoding()
...
+
+   TR::Instruction *branchInstr = generateJTYPE(TR::InstOpCode::_jal, getNode(), zero, getLabelSymbol(), cg(), self());
+   branchInstr->setEstimatedBinaryLength(RISCV_INSTRUCTION_LENGTH);
+
+   TR::Instruction *labelInstr = generateLABEL(cg(), TR::InstOpCode::label, getNode(), skipBranchLabel, branchInstr);
+   labelInstr->setEstimatedBinaryLength(0);
+
+   setLabelSymbol(skipBranchLabel);
+   setEstimatedBinaryLength(RISCV_INSTRUCTION_LENGTH);
+   //reverseLikeliness();
```
[the remark this answers is not in the document]
Comment (janvrany) [reply compiler/riscv/codegen/RVInstruction.cpp]: Thanks for looking at it! 
Agreed, removed in 01786557

File: compiler/riscv/codegen/RVInstruction.cpp [hunk 2]
```diff
@@ -411,6 +412,52 @@ uint8_t *TR::BtypeInstruction::generateBinaryEncoding()
...
+   //_farRelocation = true;
```
[the remark this answers is not in the document]
Comment (janvrany) [reply compiler/riscv/codegen/RVInstruction.cpp]: OK, for now (01786557) I removed it. I'll reintroduce it if the decision is to keep it.

File: compiler/riscv/codegen/InstOpCode.hpp [hunk 3]
```diff
@@ -47,6 +47,34 @@ class InstOpCode: public OMR::InstOpCodeConnector
...
+   /**
+    * @bried For given branch opcode, return 'reversed' opcode. For example, for BEQ return
+    * BNE, for BLT return BGE and so on. Assumes passed opcode is a branch opcode.
+    *
+    * @param[in] opcode : opcode to be reversed
+    * @return Reversed opcode
+    */
+   static TR::InstOpCode::Mnemonic reversedBranchOpCode(TR::InstOpCode::Mnemonic opcode)
+      {
+      switch (opcode) {
```
[the remark this answers is not in the document]
Comment (janvrany) [reply compiler/riscv/codegen/InstOpCode.hpp]: Argh, habits...Thanks, fixed in 01786557 (I hope)

File: compiler/riscv/codegen/OMRCodeGenerator.cpp [hunk 4]
```diff
@@ -210,6 +210,23 @@ OMR::RV::CodeGenerator::doRegisterAssignment(TR_RegisterKinds kindsToAssign)
...
    }
 
+static void
+expandFarConditionalBranches(TR::CodeGenerator *cg)
+   {
+   for (TR::Instruction *instr = cg->getFirstInstruction(); instr; instr = instr->getNext())
+      {
+      TR::BtypeInstruction *branch = instr->getBtypeInstruction();
+
+      if (branch && branch->getLabelSymbol())
```
[the remark this answers is not in the document]
Comment (janvrany) [reply compiler/riscv/codegen/OMRCodeGenerator.cpp]: Right, as the code stands, B-type instruction without label makes no sense. Removed in 01786557.

**Comment by janvrany** [comment]
@0xdaryl , @aviansie-ben How about the `_farRelocation`? Anything else to be improved?

**Comment by janvrany** [comment]
Anything else to be improved here?

**Comment by 0xdaryl** [comment]
This is a RISC-V specific change.  No need to wait for CI builds.  Merging.

**Pull request merged**